src/rust/plugin-work-queue/src/psql_queue.rs

```diff
@@ -0,0 +1,305 @@
+                last_updated = CURRENT_TIMESTAMP,
+                visible_after  = CURRENT_TIMESTAMP + INTERVAL '10 seconds'
+            FROM (
+                 SELECT execution_key, plugin_id, pipeline_message, status, creation_time, visible_after, tenant_id
+                 FROM plugin_work_queue.analyzer_plugin_executions
+                 WHERE status = 'enqueued'
+                   AND creation_time >= (CURRENT_TIMESTAMP - INTERVAL '1 day')
+                   AND (visible_after IS NULL OR visible_after <= CURRENT_TIMESTAMP)
+                 ORDER BY creation_time ASC
+                 FOR UPDATE SKIP LOCKED
+                 LIMIT 1
+             ) AS next_execution
+             WHERE plugin_work_queue.analyzer_plugin_executions.execution_key = next_execution.execution_key
+             RETURNING
+                 next_execution.execution_key as "execution_key!: ExecutionId",
+                 next_execution.plugin_id,
+                 next_execution.pipeline_message,
+                 next_execution.tenant_id
+        "#).fetch_optional(&self.pool)
+            .await?;
+
+        Ok(request.map(|request| Message { request }))
+    }
+
+    #[instrument(err)]
+    pub async fn ack_generator(
+        &self,
+        execution_key: ExecutionId,
+        status: Status,
+    ) -> Result<(), PsqlQueueError> {
+        sqlx::query!(
+            r#"
+                UPDATE plugin_work_queue.generator_plugin_executions
+                SET status = $2,
+                    last_updated = CASE
+                        WHEN status != 'processed'
+                            THEN CURRENT_TIMESTAMP
+                            ELSE last_updated
+                        END
+                WHERE execution_key = $1
+            "#,
+            execution_key.0,
+            status as _,
+        )
+        .execute(&self.pool)
+        .await?;
+        Ok(())
+    }
+
+    #[instrument(err)]
+    pub async fn ack_analyzer(
+        &self,
+        execution_key: ExecutionId,
+        status: Status,
+    ) -> Result<(), PsqlQueueError> {
+        sqlx::query!(
+            r#"
+                UPDATE plugin_work_queue.analyzer_plugin_executions
+                SET status = $2,
+                    last_updated = CASE
+                        WHEN status != 'processed'
+                            THEN CURRENT_TIMESTAMP
+                            ELSE last_updated
+                        END
+                WHERE execution_key = $1
+            "#,
+            execution_key.0,
+            status as _,
+        )
+        .execute(&self.pool)
+        .await?;
+        Ok(())
+    }
+}
+
+// Pub for testing - otherwise sqlx can't see the query
+pub async fn get_generator_status(
+    pool: &sqlx::Pool<Postgres>,
+    execution_key: &ExecutionId,
+) -> Result<Status, sqlx::Error> {
+    // The request should be marked as failed
+    let row = sqlx::query!(
+        r#"SELECT status as "status: Status"
```

Comment:
What's the deal with these column names?

Reply:
https://docs.rs/sqlx/latest/sqlx/macro.query_as.html#column-type-override-infer-from-struct-field